File: go/parsing-log-files/parsing_log_files.go
package parsinglogfiles

import (
	"regexp"
)

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^(\[TRC\]|\[DBG\]|\[INF\]|\[WRN\]|\[ERR\]|\[FTL\])`)
	return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`<(\*|~|=|-)*>`)
	return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	re := regexp.MustCompile(`(?i)".*password.*"`)
	count := 0
	for _, line := range lines {
		if re.MatchString(line) {
			count += 1
		}
	}
	return count
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`(end-of-line\d+)+`)

	return re.ReplaceAllString(text, "")
}

func TagWithUserName(lines []string) []string {
	var result []string
	re := regexp.MustCompile(`\s+User\s+(\S+)`)

	for _, line := range lines {
		match := re.FindStringSubmatch(line)
		if re.MatchString(line) {
			line = "[USR] " + match[1] + " " + line
		}
		result = append(result, line)
	}
	return result
}

File: csharp/parsing-log-files/ParsingLogFiles.cs
using System.Text.RegularExpressions;

public class LogParser
{
    public bool IsValidLine(string text)
    {
        const string pattern = @"^(\[TRC\] | \[DBG\] | \[INF\] | \[ERR\] | \[WRN\] | \[FTL\])";
        return  Regex.IsMatch(text, pattern,  RegexOptions.IgnorePatternWhitespace);
    }

    public string[] SplitLogLine(string text)
    {
        var pattern = @"<[*^=-]*>";
        return new Regex(pattern).Split(text);
    }

    public int CountQuotedPasswords(string lines)
    {
        bool[] results = new bool[lines.Length];
        var regex = new Regex(@""".*password.*""",
            RegexOptions.IgnoreCase | RegexOptions.Multiline);
        var matches = regex.Matches(lines);
        return matches.Count;
    }

    public string RemoveEndOfLineText(string line)
    {
        return Regex.Replace(line, @"end-of-line\d+", "", RegexOptions.IgnoreCase);
    }

    public string[] ListLinesWithPasswords(string[] lines)
    {
        var pattern = @".*\s(?<pw>password\S+).*";

        string[] outlines = new string[lines.Length];
        var regex = new Regex(pattern, RegexOptions.IgnoreCase);
        for (int i = 0; i < lines.Length; i++)
        {
            var matches = regex.Matches(lines[i]);
            if (matches.Count > 0)
            {
                var grps = matches[0].Groups;
                outlines[i]
                  = $"{grps["pw"].Value}: {lines[i]}";
            }
            else
            {
                outlines[i] = $"--------: {lines[i]}";
            }
        }

        return outlines;
    }
}
